### ISSUE TITLE
Consolidate on for 2^64 or 2^256 leaves in Binary and Binary Sum trees

### ISSUE THREAD

**Issue by bvrooman** [FuelLabs/fuel-merkle#38]
**Motivation**

The `fuel-merkle` library is in the process of supporting Binary, Binary Sum, and Sparse Merkle trees. All trees support address spaces of 256 bits (how this address space is broken down is out of scope for this discussion). This gives rise to the following properties:
- Each tree will have up to 2<sup>256</sup> leaves
- Each tree will have up to 2<sup>256 + 1</sup> - 1 = 2<sup>257</sup> - 1 total nodes

As the name suggests, Sparse Merkle trees have a sparse node arrangement. The leaves in a Sparse Merkle tree are positioned by the SHA256 hash of their leaf data. This means that for a Sparse Merkle tree leaf, its leaf index is equal to the hash value. The large distribution of hash values across the 256 address space results in leaves being "far apart". Leaves with no data are still a part of the tree, but act as placeholders. Traversing the tree from root to leaf is done by travelling from the root to the leaf hash.

Recall that a tree may have up to 2<sup>257</sup> - 1 total nodes. Because the address space of sparse nodes is restricted to 256 bits, there can be up to 2<sup>256</sup> unique addresses. This means that collisions are theoretically possible; however, this address space is so large that collisions are impossible in practice.

The sparse nature of the SMT is in contrast to the "compact" node arrangement created by the Binary and Binary Sum trees. Rather than placing leaves sparsely and using placeholders in between them, these trees treat leaves as a continuous list; leaves are indexed based on the order in which they are pushed to the tree, starting at 0 and incrementing by 1 at each push. This means the leaf index is independent of the leaf data (and by extension the leaf hash value), and is calculated solely through monotonically incrementing the tree's leaf count. This distinction is especially important when traversing the tree from root to leaf; when traversing a BMT or BMST, we must traverse from the root to the leaf index - not the root hash. 

One consequence of this bookkeeping is that we require an integer (or an integer-like structure) that keeps track of our leaf count. In order to truly support our 256 bit address space, we must use an integer that supports up a value up to 2<sup>256</sup> - 1, i.e. `u256`. In Rust, such an integer is not natively supported: we are restricted to types `u64` or `u128`. Furthermore, some traversal algorithms in the BMT and BMST may require that internal nodes be indexed as well. The use of such algorithms necessitate support for up to 2<sup>257</sup> - 1 indices, i.e. `u512`. Again, such a type is not natively supported.

**Options**

1. Specify that the Binary Merkle tree and Binary Merkle Sum trees support only up to 2<sup>64</sup> leaves. In practical terms, this space is very large and may already be sufficient.
2. Add support for 256 bit and 512 bit integers to allow the Binary Merkle tree and Binary Merkle Sum trees to support 2<sup>256</sup> leaves.

**TL;DR**

- SMTs support 2<sup>256</sup> leaves by using 32 bit hashes as leaf keys 
- BMTs and BMSTs do not natively support 2<sup>256</sup> leaves because this requires 256 bit integer indices
- We can either reduce the number of supported leaves to 2<sup>64</sup> for these trees or support 256 bit integers
